File: trunk/Base Libraries/Base Render Library/Camera Control/Input/Mouse.cs
/*
 * Author: Bogolepov Denis (denisbogol@sandy.ru)
 */
 
using System;
using Base.MathTools;

namespace Base.Graphics
{
	public class Mouse
	{
		#region Private Fields
		
		private int locationX = 0;
		
		private int locationY = 0;

		#endregion		
		
		#region Public Fields
		
		public Camera Camera = null;
		
		public float Step = 0.005f;
		
		#endregion
		
		#region Constructor
		
		public Mouse(Camera camera)
		{
			Camera = camera;
		}		
		
		public Mouse(Camera camera, float speed)
			: this(camera)
		{
			Step = speed;
		}
		
		#endregion
			
		#region Public Methods
		
		public void OnStateChange(int x, int y)
		{
			locationX = x;
			
			locationY = y;			
		}
		
		public void OnMouseMove(int x, int y)
		{
			if (null == Camera)
				return;
												
        	if (locationX != x)
        	{
        		Camera.Rotate(Vector3.UnitZ * (x - locationX) * Step);
        	}
        	
        	if (locationY != y)
        	{
        		Camera.Rotate(Vector3.UnitY * (y - locationY) * Step);
			}
			
			locationX = x;
			
			locationY = y;
		}
		
		#endregion
	}
}

File: trunk/Source/Base Libraries/Base Render Library/Mouse.cpp
/*
 * Author: Denis Bogolepov  ( denisbogol@sandy.ru )
 */

#include "Mouse.h"

#include <GL/glfw.h>

namespace Render
{
	//--------------------------------------- Constructors ----------------------------------------
	
	Mouse :: Mouse ( float step )
	{
		Step = step;
	}

	//-------------------------------------- Event Handlers ---------------------------------------
	
	void Mouse :: StateChange ( int state )
	{
		glfwGetMousePos ( &LocationX, &LocationY );

		Active = state > 0;
	}
	
	void Mouse :: MouseMove ( int x, int y )
	{
		if ( Active )
		{
			if ( LocationX != x )
			{
				DeltaX += x - LocationX;

				LocationX = x;
			}
			
			if ( LocationY != y )
			{
				DeltaY += y - LocationY;

				LocationY = y;
			}					
		}
	}

	//--------------------------------------- Camera Change ---------------------------------------

	void Mouse :: Apply ( Camera * camera )
	{
		if ( DeltaX != 0 )
		{
			camera->RotateLocal ( -DeltaX * Step, Vector3D :: AxisY );

			DeltaX = 0;
		}
		
		if ( DeltaY != 0 )
		{
			camera->RotateLocal ( DeltaY * Step, Vector3D :: AxisX );

			DeltaY = 0;
		}
	}
}